clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/agents/TitusStreamingUpdateAgent.java

```diff
@@ -185,37 +187,73 @@ public void executeAgent(Agent agent) {
             .build()
         );
+
+        Boolean snapshotComplete = false;
+        Boolean savedSnapshot = false;
+
         Long startTime = System.currentTimeMillis();
-        while (notificationIt.hasNext()) {
-          JobChangeNotification notification = notificationIt.next();
-          switch (notification.getNotificationCase()) {
-            case JOBUPDATE:
-              updateJob(notification.getJobUpdate().getJob());
-              break;
-            case TASKUPDATE:
-              updateTask(notification.getTaskUpdate().getTask());
-              break;
-            case SNAPSHOTEND:
-              lastUpdate.set(0); // last update should already be initialized to 0 but this forces an update regardless
-              log.info("Snapshot finished {}", getAgentType());
-              tasks.keySet().retainAll(jobs.keySet());
-              break;
+
+        while (continueStreaming(startTime)) {
+          try {
+            while (notificationIt.hasNext() && continueStreaming(startTime)) {
+              JobChangeNotification notification = notificationIt.next();
+              switch (notification.getNotificationCase()) {
+                case JOBUPDATE:
+                  updateJob(notification.getJobUpdate().getJob());
+                  break;
+                case TASKUPDATE:
+                  updateTask(notification.getTaskUpdate().getTask(), snapshotComplete);
+                  break;
+                case SNAPSHOTEND:
+                  lastUpdate.set(0); // last update should already be initialized to 0 but this forces it
+                  log.info("{} snapshot finished in {}ms", getAgentType(), System.currentTimeMillis() - startTime);
+                  tasks.keySet().retainAll(jobs.keySet());
+                  if (snapshotComplete) {
+                    log.error("{} received >1 SNAPSHOTEND events, this is unexpected and may be handled incorrectly",
+                      getAgentType()
+                    );
+                  }
+                  snapshotComplete = true;
+                  break;
+              }
+
+              if (snapshotComplete) {
+                writeToCache(!savedSnapshot);
+                if (!savedSnapshot) {
+                  savedSnapshot = true;
+                }
+              }
+            }
+          } catch (io.grpc.StatusRuntimeException e) {
+            log.warn("gRPC exception while streaming {} updates, attempting to reconnect", getAgentType(), e);
+            notificationIt = titusClient.observeJobs(
+              ObserveJobsQuery.newBuilder()
+                .putFilteringCriteria("jobType", "SERVICE")
+                .putFilteringCriteria("attributes", "source:spinnaker")
+                .build()
+            );
+            snapshotComplete = false;
+            savedSnapshot = false;
+          } catch (Exception e) {
+            log.error("Exception while streaming {} titus updates", getAgentType(), e);
```

Comment:
This exposed 2 NPE's in the task and job update processing logic that resulted in premature agent termination.